### PR TITLE
Allow for better customization of dropdowns

### PR DESCRIPTION
**Pull request by SirCmpwn**
Now more closely matches the model of checkboxes and sliders, where the framework user is expected to provide their own implementation derived from abstract classes.

### CODE REVIEW

File: osu.Framework/Graphics/UserInterface/DropDownComboBox.cs [hunk 1]
```diff
@@ -9,14 +9,16 @@
...
         protected Box Background;
         protected virtual Color4 BackgroundColour => Color4.DarkGray;
         protected virtual Color4 BackgroundColourHover => Color4.Gray;
         protected Container Foreground;
-        protected Drawable Label;
-        protected Drawable Caret;
+
+        protected override Container<Drawable> Content => Foreground;
+
+        protected internal abstract string Label { get; set; }
```
Comment (SirCmpwn) [inline osu.Framework/Graphics/UserInterface/DropDownComboBox.cs]: Thought about changing DropDownComboBox to DropDownComboBox<T> and making this Value or DisplayValue or something. Is probably the better design but for our use case it doesn't really matter. Feedback welcome.

File: osu.Framework/Graphics/UserInterface/DropDownMenu.cs [hunk 2]
```diff
@@ -51,10 +51,7 @@ public IEnumerable<DropDownMenuItem<T>> Items
         protected Box DropDownBackground;
         protected virtual float DropDownListSpacing => 0;
 
-        protected virtual DropDownComboBox CreateComboBox()
-        {
-            return new DropDownComboBox();
-        }
+        protected abstract DropDownComboBox CreateComboBox();
```
Comment (SirCmpwn) [inline osu.Framework/Graphics/UserInterface/DropDownMenu.cs]: I'm not a fan of this having to be a method (this is left over from the previous implementation of DropBoxes) but I'm not really sure what a better solution is. Can't pass it into base() cleanly.

File: osu.Framework.VisualTests/Tests/TestCaseDropDownBox.cs [hunk 3]
```diff
@@ -190,61 +124,22 @@ protected override void OnSelectChange()
                     return;
 
                 FormatBackground();
-                FormatCaret();
-                FormatLabel();
+                formatCaret();
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseDropDownBox.cs]: is there any reason to leave this stuff here? seems unnecessary and ugly (the selected item already gets highlighted with a different colour background).

File: osu.Framework.VisualTests/Tests/TestCaseDropDownBox.cs [hunk 4]
```diff
@@ -169,19 +79,43 @@ private class StyledDropDownComboBox : DropDownComboBox
...
+                    label = new SpriteText(),
+                };
             }
         }
 
         private class StyledDropDownMenuItem : DropDownMenuItem<string>
         {
+            private SpriteText caret, label;
+        
             public StyledDropDownMenuItem(string text) : base(text, text)
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseDropDownBox.cs]: man, even as a test case this is ugly stuff (i realise it wasn't your addition).

File: osu.Framework/Graphics/UserInterface/DropDownComboBox.cs (hunk 1 above)
Comment (SirCmpwn) [inline osu.Framework/Graphics/UserInterface/DropDownComboBox.cs]: Explored this, is not necessary or practical.

**Pull request merged**